### PR TITLE
[BE] Move tests using functional API to OptimizerInfo

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123139
* #123134
* __->__ #122822

### CODE REVIEW

**Review by albanD — APPROVED** [review]
SGTM

File: torch/testing/_internal/common_optimizers.py [hunk 1]
```diff
@@ -1071,6 +1071,13 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "CompiledOptimizerParityTests",
                 "test_correctness",
             ),
+            DecorateInfo(
+                skipIfTorchDynamo(
+                    "This test uses mocks, which dynamo does not support"
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1076]: The error trace if you're interested is:
```
======================================================================
FAIL: test_defaults_changed_to_foreach_SGD_cuda_float32 (__main__.TestOptimRenewedCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2741, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2741, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_device_type.py", line 419, in instantiated_test
    result = test(self, **param_kwargs)
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_optimizers.py", line 186, in test_wrapper
    return test(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_device_type.py", line 1166, in only_fn
    return fn(slf, *args, **kwargs)
  File "/data/users/janeyx/pytorch/test/test_optim.py", line 1317, in test_defaults_changed_to_foreach
    optim.step()
  File "/data/users/janeyx/pytorch/torch/optim/optimizer.py", line 391, in wrapper
    out = func(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 939, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state, skip=1)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 802, in _convert_frame
    result = inner_convert(
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 400, in _convert_frame_assert
    return _compile(
  File "/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 686, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/data/users/janeyx/pytorch/torch/_dynamo/utils.py", line 265, in time_wrapper
    r = func(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 541, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/data/users/janeyx/pytorch/torch/_dynamo/bytecode_transformation.py", line 1036, in transform_code_object
    transformations(instructions, code_options)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 165, in _fn
    return fn(*args, **kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/convert_frame.py", line 503, in transform
    tracer.run()
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 2116, in run
    super().run()
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 839, in run
    while self.step():
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 753, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 481, in wrapper
    return inner_fn(self, inst)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 1255, in CALL_FUNCTION_KW
    self.call_function(fn, args, kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 693, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/data/users/janeyx/pytorch/torch/_dynamo/variables/functions.py", line 295, in call_function
    return super().call_function(tx, args, kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/variables/functions.py", line 90, in call_function
    return tx.inline_user_function_return(
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 699, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 2263, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 2377, in inline_call_
    tracer.run()
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 839, in run
    while self.step():
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 753, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 481, in wrapper
    return inner_fn(self, inst)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 1255, in CALL_FUNCTION_KW
    self.call_function(fn, args, kwargs)
  File "/data/users/janeyx/pytorch/torch/_dynamo/symbolic_convert.py", line 692, in call_function
    raise AssertionError(f"Attempt to trace forbidden callable {inner_fn}")
AssertionError: Attempt to trace forbidden callable <MagicMock name='_multi_tensor_sgd' id='140443477916208'>
```

**Comment by janeyx99** [comment]
@pytorchbot merge -i

failure is a known flaky test

**Pull request closed without merging**